File: android/src/main/kotlin/com/giphy/giphy_flutter_sdk/GiphyFlutterMediaView.kt
package com.giphy.giphy_flutter_sdk

import android.annotation.SuppressLint
import android.content.Context
import android.util.Log
import android.view.View
import com.facebook.drawee.drawable.ScalingUtils
import com.giphy.giphy_flutter_sdk.dto.GiphyFlutterResizeMode
import com.giphy.giphy_flutter_sdk.dto.GiphyApiException
import com.giphy.giphy_flutter_sdk.dto.toMedia
import com.giphy.giphy_flutter_sdk.dto.toRenditionType
import com.giphy.giphy_flutter_sdk.utils.require
import com.giphy.sdk.core.GPHCore
import com.giphy.sdk.core.models.Media
import com.giphy.sdk.core.models.enums.RenditionType
import com.giphy.sdk.ui.utils.aspectRatio
import com.giphy.sdk.ui.views.GPHMediaView
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.platform.PlatformView

internal class GiphyFlutterMediaView(
    context: Context?,
    messenger: BinaryMessenger,
    viewId: Int,
    params: Map<String, Any>?
) : PlatformView {
    private val view: GPHMediaView
    private val channel = MethodChannel(messenger, "com.giphyfluttersdk/mediaView$viewId")

    private var loadedMedia: Media? = null
    private var autoPlay: Boolean = true
    private var renditionType = RenditionType.fixedWidth
    private var resizeMode = GiphyFlutterResizeMode.DEFAULT_MODE

    init {
        view = GPHMediaView(context.require("Context must not be null"))
        channel.setMethodCallHandler { call, result ->
            when (call.method) {
                "setMediaId" -> {
                    setMediaWithId(call.argument<String>("mediaId"))
                    result.success(null)
                }

                "setMedia" -> {
                    setMedia(call.argument<HashMap<String, Any?>>("media")?.toMedia())
                    result.success(null)
                }

                "setAutoPlay" -> {
                    setAutoPlay(call.argument<Boolean?>("autoPlay"))
                    result.success(null)
                }

                "setRenditionType" -> {
                    setRenditionType(call.arguments as? HashMap<String, Any?>)
                    result.success(null)
                }

                "setResizeMode" -> {
                    setResizeMode(call.argument<String?>("resizeMode"))
                    result.success(null)
                }

                "setShowCheckeredBackground" -> {
                    setShowCheckeredBackground(call.argument<Boolean?>("showCheckeredBackground"))
                    result.success(null)
                }

                "pause" -> {
                    view.pause()
                    result.success(null)
                }

                "resume" -> {
                    view.play()
                    result.success(null)
                }

                else -> result.notImplemented()
            }
        }
    }

    private fun setAutoPlay(value: Boolean?) {
        if (value != null) {
            autoPlay = value
        }
    }

    @SuppressLint("LongLogTag")
    private fun setMediaWithId(mediaId: String?) {
        if (mediaId == null) {
            return
        }

        GPHCore.gifById(mediaId) { result, e ->
            loadedMedia = result?.data
            syncMedia()
            e?.let {
                val exception = GiphyApiException(it)
                val errorMessage = exception.errorMessage
                channel.invokeMethod(
                    "onError", mapOf(
                        "error" to exception.errorMessage
                    )
                )
            }
        }
    }

    private fun setMedia(media: Media?) {
        loadedMedia = media
        syncMedia()
    }

    private fun setRenditionType(map: HashMap<String, Any?>?) {
        renditionType = map?.toRenditionType() ?: RenditionType.fixedWidth
        syncMedia()
    }

    private fun setResizeMode(value: String?) {
        resizeMode = GiphyFlutterResizeMode.fromString(value) ?: GiphyFlutterResizeMode.DEFAULT_MODE
        adjustResizeMode()
        syncMedia()
    }

    private fun setShowCheckeredBackground(value: Boolean?) {
        view.isBackgroundVisible = value ?: true
        syncMedia()
    }

    private fun syncMedia() {
        if (loadedMedia != null) {
            view.setMedia(loadedMedia, renditionType)
        }
        view.aspectRatio = loadedMedia?.aspectRatio ?: view.aspectRatio
        if (!autoPlay) {
            view.pause()
        }
    }

    private fun adjustResizeMode() {
        when (resizeMode) {
            GiphyFlutterResizeMode.CENTER -> {
                view.scaleType = ScalingUtils.ScaleType.CENTER_INSIDE
            }

            GiphyFlutterResizeMode.CONTAIN -> {
                view.scaleType = ScalingUtils.ScaleType.FIT_CENTER
            }

            GiphyFlutterResizeMode.COVER -> {
                view.scaleType = ScalingUtils.ScaleType.CENTER_CROP
            }

            GiphyFlutterResizeMode.STRETCH -> {
                view.scaleType = ScalingUtils.ScaleType.FIT_XY
            }
        }
    }

    override fun getView(): View {
        return view
    }

    override fun dispose() {
        view.recycle()
        channel.setMethodCallHandler(null)
    }

}

File: ios/giphy_flutter_sdk/Sources/giphy_flutter_sdk/GiphyFlutterMediaView.swift
//
//  GiphyFlutterMediaView.swift
//  giphy_flutter_sdk
//
//  Created by Alex on 27.2.2024.
//

import Flutter
import UIKit
import GiphyUISDK

public class GiphyFlutterMediaView: NSObject, FlutterPlatformView {
    private let mediaView: GPHMediaView = GPHMediaView()
    
    private static let DEFAULT_RENDITION_TYPE: GPHRenditionType = .fixedWidth
    
    private var channel: FlutterMethodChannel?
    
    private var media: GPHMedia? {
        didSet {
            syncMediaViewSource()
        }
    }
    
    private var renditionType: GPHRenditionType = DEFAULT_RENDITION_TYPE {
        didSet {
            syncMediaViewSource()
        }
    }
    
    private var resizeMode: GiphyFlutterResizeMode = .defaultMode {
        didSet {
            adjustResizeMode()
        }
    }
    
    init(frame: CGRect, viewId: Int64, args: Any?, binaryMessenger messenger: FlutterBinaryMessenger) {
        super.init()
        setupView(frame: frame)
        adjustResizeMode()
        createMethodChannel(viewId: viewId, messenger: messenger)
    }
    
    deinit {
        channel?.setMethodCallHandler(nil)
    }
    
    public func view() -> UIView {
        return mediaView
    }
    
    
    private func pause() {
        mediaView.stopAnimating()
    }
    
    private func resume() {
        mediaView.startAnimating()
    }
    
    private func setMediaWithId(_ value: String?) -> Void {
        guard
            let mediaId = value
        else {
            media = nil
            return
        }
        
        GiphyCore.shared.gifByID(mediaId) { [weak self] (response, error) in
            guard let self = self else {
                return
            }
            if let error = error {
                channel?.invokeMethod("onError", arguments: [
                    "error": error.localizedDescription
                ])
            } else {
                self.media = response?.data
            }
        }
    }
    
    private func setAutoPlay(_ value: Bool) -> Void {
        mediaView.autoPlayAnimatedImage = value
    }
    
    private func setRenditionType(_ value: String?) -> Void {
        renditionType = GPHRenditionType.fromString(value) ?? GiphyFlutterMediaView.DEFAULT_RENDITION_TYPE
    }
    
    private func setResizeMode(_ value: String?) -> Void {
        resizeMode = GiphyFlutterResizeMode.fromString(value: value) ?? .defaultMode
    }
    
    private func createMethodChannel(viewId: Int64, messenger: FlutterBinaryMessenger) {
        channel = FlutterMethodChannel(name: "com.giphyfluttersdk/mediaView\(viewId)", binaryMessenger: messenger)
        channel?.setMethodCallHandler({
            [weak self] (call: FlutterMethodCall, result: @escaping FlutterResult) -> Void in
            guard let self = self else {
                result(FlutterError(code: "SELF_NIL", message: "Reference to self is nil", details: nil))
                return
            }
            
            let arguments = call.arguments as? [String: Any?] ?? [:]
            
            switch call.method {
            case "setMediaId":
                if let mediaId = arguments["mediaId"] as? String {
                    self.setMediaWithId(mediaId)
                    result(nil)
                } else {
                    result(FlutterError(code: "INVALID_ARGUMENTS", message: "Expected a mediaId", details: nil))
                }
                
            case "setMedia":
                if let mediaDict = arguments["media"] as? [String: Any?] {
                    self.media = mediaDict.toMedia()
                    result(nil)
                } else {
                    result(FlutterError(code: "INVALID_ARGUMENTS", message: "Expected a media dictionary", details: nil))
                }
                
            case "setAutoPlay":
                if let autoPlay = arguments["autoPlay"] as? Bool {
                    self.setAutoPlay(autoPlay)
                    result(nil)
                } else {
                    result(FlutterError(code: "INVALID_ARGUMENTS", message: "Expected a boolean for autoPlay", details: nil))
                }
                
            case "setRenditionType":
                if let renditionType = arguments["renditionType"] as? String {
                    self.setRenditionType(renditionType)
                    result(nil)
                } else {
                    result(FlutterError(code: "INVALID_ARGUMENTS", message: "Expected a map for renditionType", details: nil))
                }
                
            case "setResizeMode":
                if let resizeMode = arguments["resizeMode"] as? String {
                    self.setResizeMode(resizeMode)
                    result(nil)
                } else {
                    result(FlutterError(code: "INVALID_ARGUMENTS", message: "Expected a string for resizeMode", details: nil))
                }
                
            case "setShowCheckeredBackground":
                // To maintain consistency with Android, do not return an error, even though it's not supported.
                result(nil)
                
            case "pause":
                self.mediaView.stopAnimating()
                result(nil)
                
            case "resume":
                self.mediaView.startAnimating()
                result(nil)
                
            default:
                result(FlutterMethodNotImplemented)
            }            
        })
    }
    
    private func setupView(frame: CGRect) -> Void {
        mediaView.frame = frame
        mediaView.autoresizingMask = [.flexibleWidth, .flexibleHeight]
        mediaView.clipsToBounds = true
    }
    
    private func adjustResizeMode() -> Void {
        switch resizeMode {
        case .center:
            mediaView.contentMode = .center
        case .contain:
            mediaView.contentMode = .scaleAspectFit
        case .cover:
            mediaView.contentMode = .scaleAspectFill
        case .stretch:
            mediaView.contentMode = .scaleToFill
        }
    }
    
    private func syncMediaViewSource() -> Void {
        DispatchQueue.main.async { [weak self] in
            guard let self = self else {
                return
            }
            if self.media == nil {
                self.mediaView.media = self.media
            } else {
                self.mediaView.setMedia(self.media!, rendition: self.renditionType)
            }
        }
    }
}
